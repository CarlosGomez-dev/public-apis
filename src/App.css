.App {
  max-width: 1200px;
  margin: 0 auto;
  padding: 2rem;
}

a[target='_blank']::after {
  content: url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAoAAAAKCAYAAACNMs+9AAAAQElEQVR42qXKwQkAIAxDUUdxtO6/RBQkQZvSi8I/pL4BoGw/XPkh4XigPmsUgh0626AjRsgxHTkUThsG2T/sIlzdTsp52kSS1wAAAABJRU5ErkJggg==);
  margin: 0 5px 0 3px;
}
