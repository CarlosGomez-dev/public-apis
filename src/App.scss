@import './colors';

.main-page {
  width: clamp(400px, 80%, 1000px);
  margin: 0 auto;
  padding: 2rem;
  margin-top: 3rem;
}

.page-title {
  position: fixed;
  display: flex;
  place-items: center;
  place-content: center;
  top: 0;
  left: 0;
  z-index: 1;
  padding: 1rem;
  background-color: $background-color;
  text-align: center;
  width: 100%;
  width: 100vw;
  min-width: 400px;
}
.page-error,
.page-no-results,
.page-loading {
  text-align: center;
  padding: 1rem;
  font-size: 1.3rem;
}
