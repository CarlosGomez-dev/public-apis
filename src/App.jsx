import { memo, useCallback, useEffect, useRef, useState } from 'react';
import { DataList, FilterForm, Pagination } from './components';
import { useFetchPublicApi } from './hooks';
import { sortObjectArray } from './utils';
import plugIn from './assets/plug-in.svg';
import './App.scss';

const FilterFormMemo = memo(FilterForm);

export const App = () => {
  const [apiData, isLoading, error] = useFetchPublicApi('entries');
  const [{ sortBy, ascending }, setSortBy] = useState({ sortBy: '', ascending: '' });

  const [categoryFilter, setCategoryFilter] = useState('');
  const [searchFilter, setSearchFilter] = useState('');
  const [currentPage, setCurrentPage] = useState(1);
  const [pageLimit, setPageLimit] = useState(20);
  const inputRef = useRef(null);

  const sortedData = sortObjectArray(apiData, sortBy, ascending);
  const categoryApiData = categoryFilter
    ? sortedData.filter(api => api.Category === categoryFilter)
    : sortedData;
  const filteredApiData = categoryApiData.filter(api =>
    (api.API + api.Description).toLowerCase().includes(searchFilter.toLowerCase()),
  );
  const paginatedApiData = filteredApiData.slice(
    pageLimit * (currentPage - 1),
    pageLimit * currentPage,
  );

  const handleSearch = useCallback(searchText => {
    setCurrentPage(1);
    setSearchFilter(searchText);
  }, []);

  const handleCategory = useCallback(category => {
    setCurrentPage(1);
    setCategoryFilter(category);
  }, []);

  const handlePageLimit = useCallback(newLimit => {
    setCurrentPage(1);
    setPageLimit(newLimit);
  }, []);

  const handleSort = useCallback(
    sortColumn => {
      if (ascending === 'desc') {
        return setSortBy({ sortBy: '', ascending: '' });
      }
      if (sortColumn === sortBy) {
        return setSortBy({ sortBy: sortColumn, ascending: 'desc' });
      }
      setSortBy({ sortBy: sortColumn, ascending: 'asc' });
    },
    [ascending, sortBy],
  );

  const handleKeyUp = event => {
    if (event.key === '/') {
      event.preventDefault();
      inputRef.current.focus();
      inputRef.current.select();
    }
  };

  useEffect(() => {
    inputRef.current.focus();
    document.body.addEventListener('keyup', handleKeyUp);
    return () => document.body.removeEventListener('keyup', handleKeyUp);
  }, []);

  return (
    <div className='main-page'>
      <header>
        <h1 className='page-title'>
          <img src={plugIn} alt='plug in' />
          <span>Public APIs</span>
        </h1>
        <FilterFormMemo
          inputRef={inputRef}
          pageLimit={pageLimit}
          handleSearch={handleSearch}
          handlePageLimit={handlePageLimit}
          handleSort={handleSort}
          handleCategory={handleCategory}
        />
      </header>
      <Pagination
        totalResults={filteredApiData.length}
        currentPage={currentPage}
        pageLimit={pageLimit}
        setCurrentPage={setCurrentPage}
      />
      {error && (
        <p className='page-error'>There was an error fetching the API. Please try again later.</p>
      )}
      {isLoading && <p className='page-loading'>Loading results...</p>}
      {!isLoading && paginatedApiData.length === 0 && (
        <p className='page-no-results'>
          No results, try another search term or change the filters to see more results.
        </p>
      )}
      {!isLoading && paginatedApiData.length > 0 && (
        <DataList apiData={paginatedApiData} handleSort={handleSort} />
      )}
    </div>
  );
};
