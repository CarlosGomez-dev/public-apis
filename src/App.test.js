import { act, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { App } from './App';

const options = [
  ['results per page', '50', '50 / page'],
  ['sort by', 'API', 'Name'],
  ['categories', 'Music', 'Music'],
];

const categories = [
  'Data Validation',
  'Development',
  'Dictionaries',
  'Documents \u0026 Productivity',
  'Environment',
  'Events',
  'Finance',
  'Food \u0026 Drink',
  'Games \u0026 Comics',
  'Geocoding',
  'Government',
  'Health',
  'Jobs',
  'Machine Learning',
  'Music',
];

const entries = [
  {
    API: 'IUCN',
    Description: 'IUCN Red List of Threatened Species',
    Auth: 'apiKey',
    HTTPS: false,
    Cors: 'unknown',
    Link: 'http://apiv3.iucnredlist.org/api/v3/docs',
    Category: 'Animals',
  },
  {
    API: 'PageCDN',
    Description: 'Public API for javascript, css and font libraries on PageCDN',
    Auth: 'apiKey',
    HTTPS: true,
    Cors: 'yes',
    Link: 'https://pagecdn.com/docs/public-api',
    Category: 'Development',
  },
  {
    API: '7digital',
    Description: 'Api of Music store 7digital',
    Auth: 'OAuth',
    HTTPS: true,
    Cors: 'unknown',
    Link: 'https://docs.7digital.com/reference',
    Category: 'Music',
  },
];

let mockError = null;
jest.mock('./hooks', () => ({
  useFetchPublicApi: endpoint => {
    const data = endpoint === 'categories' ? categories : entries;
    const isLoading = false;
    return [data, isLoading, mockError];
  },
}));

describe('App', () => {
  it('matches snapshot', () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });
  it('renders results on page load', () => {
    render(<App />);
    expect(screen.getAllByRole('row').length).toBeGreaterThan(1);
  });
  it('focuses search input on page load', () => {
    render(<App />);
    expect(screen.getByLabelText(/search/i)).toHaveFocus();
  });
  it('allows users to tab away and return to input', () => {
    const { container } = render(<App />);
    userEvent.tab();
    userEvent.tab();
    userEvent.tab();
    expect(screen.getByLabelText(/search/i)).not.toHaveFocus();
    userEvent.type(container, '/');
    expect(screen.getByLabelText(/search/i)).toHaveFocus();
  });
  it.each(options)('contains filter option: %s', async (label, option, displayValue) => {
    render(<App />);
    userEvent.selectOptions(screen.getByLabelText(new RegExp(label, 'i')), option);
    screen.getByDisplayValue(displayValue);
  });
  it('contains search input to filter results', async () => {
    render(<App />);
    const header = 1;
    const inputText = 'this returns no results';
    const input = screen.getByLabelText(/search/i);
    expect(screen.getByRole('table')).toBeInTheDocument();
    expect(screen.getAllByRole('row').length).toBe(entries.length + header);
    jest.useFakeTimers();
    userEvent.type(input, inputText);
    screen.getByDisplayValue(inputText);
    act(() => jest.runAllTimers());
    expect(screen.queryByRole('table')).not.toBeInTheDocument();
  });
  it('contains allows sorting with headers', () => {
    render(<App />);
    const headers = screen.getAllByRole('columnheader');
    const tableRows = screen.getAllByRole('row')[1];
    userEvent.click(headers[0].firstElementChild);
    expect(tableRows).not.toEqual(screen.getAllByRole('row')[1]);
    userEvent.click(headers[0].firstElementChild);
    expect(tableRows).not.toEqual(screen.getAllByRole('row')[1]);
    userEvent.click(headers[0].firstElementChild);
    expect(tableRows).toEqual(screen.getAllByRole('row')[1]);
  });
  it('displays an error message if data fetch failed', async () => {
    mockError = { error: 'Failed' };
    render(<App />);
    await screen.findByText(/error/i);
    mockError = null;
  });
});
