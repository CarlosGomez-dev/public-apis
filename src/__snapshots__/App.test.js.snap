// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches snapshot 1`] = `
<div>
  <div
    class="main-page"
  >
    <header>
      <h1
        class="page-title"
      >
        <img
          alt="plug in"
          src="plug-in.svg"
        />
        <span>
          Public APIs
        </span>
      </h1>
      <form
        class="form"
      >
        <label
          class="label"
          for="filter"
        >
          Search
        </label>
        <input
          class="input search"
          id="filter"
          placeholder="e.g.: Weather"
          type="search"
          value=""
        />
        <span>
          Press / to jump to the search box
        </span>
        <section
          class="form-filters"
        >
          <div
            class="form-select "
          >
            <label
              class="label"
              for="Results-per-Page"
            >
              Results per Page
            </label>
            <select
              class="select"
              id="Results-per-Page"
            >
              <option
                value="10"
              >
                10 / page
              </option>
              <option
                selected=""
                value="20"
              >
                20 / page
              </option>
              <option
                value="50"
              >
                50 / page
              </option>
              <option
                value="100"
              >
                100 / page
              </option>
            </select>
          </div>
          <div
            class="form-select sort-by-select"
          >
            <label
              class="label"
              for="Sort-by"
            >
              Sort by
            </label>
            <select
              class="select"
              id="Sort-by"
            >
              <option
                value=""
              >
                -
              </option>
              <option
                value="API"
              >
                Name
              </option>
              <option
                value="Description"
              >
                Description
              </option>
              <option
                value="Auth"
              >
                Auth
              </option>
              <option
                value="HTTPS"
              >
                HTTPS
              </option>
              <option
                value="Category"
              >
                Category
              </option>
            </select>
          </div>
          <div
            class="form-select "
          >
            <label
              class="label"
              for="Categories"
            >
              Categories
            </label>
            <select
              class="select"
              id="Categories"
            >
              <option
                value=""
              >
                -
              </option>
              <option
                value="Data Validation"
              >
                Data Validation
              </option>
              <option
                value="Development"
              >
                Development
              </option>
              <option
                value="Dictionaries"
              >
                Dictionaries
              </option>
              <option
                value="Documents & Productivity"
              >
                Documents & Productivity
              </option>
              <option
                value="Environment"
              >
                Environment
              </option>
              <option
                value="Events"
              >
                Events
              </option>
              <option
                value="Finance"
              >
                Finance
              </option>
              <option
                value="Food & Drink"
              >
                Food & Drink
              </option>
              <option
                value="Games & Comics"
              >
                Games & Comics
              </option>
              <option
                value="Geocoding"
              >
                Geocoding
              </option>
              <option
                value="Government"
              >
                Government
              </option>
              <option
                value="Health"
              >
                Health
              </option>
              <option
                value="Jobs"
              >
                Jobs
              </option>
              <option
                value="Machine Learning"
              >
                Machine Learning
              </option>
              <option
                value="Music"
              >
                Music
              </option>
            </select>
          </div>
        </section>
      </form>
    </header>
    <nav
      class="pagination"
    >
      <p>
        3
         results
      </p>
      <div
        class="button-container"
      >
        <button
          disabled=""
        >
          &lt;&lt;
        </button>
        <button
          disabled=""
        >
          Prev
        </button>
        <button
          disabled=""
        >
          Next
        </button>
        <button
          disabled=""
        >
          &gt;&gt;
        </button>
      </div>
      <p>
        Page 
        1
         of 
        1
      </p>
    </nav>
    <table
      class="api-table"
    >
      <thead>
        <tr>
          <th>
            <button
              class="api-table-header"
            >
              Name
            </button>
          </th>
          <th>
            <button
              class="api-table-header"
            >
              Description
            </button>
          </th>
          <th>
            <button
              class="api-table-header"
            >
              Auth
            </button>
          </th>
          <th>
            <button
              class="api-table-header"
            >
              HTTPS
            </button>
          </th>
          <th>
            <button
              class="api-table-header"
            >
              Category
            </button>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="api-name"
          >
            <a
              href="http://apiv3.iucnredlist.org/api/v3/docs"
              rel="noopener noreferrer"
              target="_blank"
            >
              <span>
                IUCN
              </span>
            </a>
          </td>
          <td
            class="api-description"
          >
            <span>
              IUCN Red List of Threatened Species
            </span>
          </td>
          <td>
            <span
              class="badge auth"
            >
              apiKey
            </span>
          </td>
          <td>
            <span
              class="badge no-https"
            >
              HTTP
              
            </span>
          </td>
          <td>
            <span
              class="badge category"
            >
              Animals
            </span>
          </td>
        </tr>
        <tr>
          <td
            class="api-name"
          >
            <a
              href="https://pagecdn.com/docs/public-api"
              rel="noopener noreferrer"
              target="_blank"
            >
              <span>
                PageCDN
              </span>
            </a>
          </td>
          <td
            class="api-description"
          >
            <span>
              Public API for javascript, css and font libraries on PageCDN
            </span>
          </td>
          <td>
            <span
              class="badge auth"
            >
              apiKey
            </span>
          </td>
          <td>
            <span
              class="badge https"
            >
              HTTP
              S
            </span>
          </td>
          <td>
            <span
              class="badge category"
            >
              Development
            </span>
          </td>
        </tr>
        <tr>
          <td
            class="api-name"
          >
            <a
              href="https://docs.7digital.com/reference"
              rel="noopener noreferrer"
              target="_blank"
            >
              <span>
                7digital
              </span>
            </a>
          </td>
          <td
            class="api-description"
          >
            <span>
              Api of Music store 7digital
            </span>
          </td>
          <td>
            <span
              class="badge auth"
            >
              OAuth
            </span>
          </td>
          <td>
            <span
              class="badge https"
            >
              HTTP
              S
            </span>
          </td>
          <td>
            <span
              class="badge category"
            >
              Music
            </span>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
